### PR TITLE
[SPARK-27275][CORE] Fix potential corruption in EncryptedMessage.transferTo (2.4)

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Backport https://github.com/apache/spark/pull/24211 to 2.4

## How was this patch tested?

Jenkins

### CODE REVIEW

File: common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java [hunk 1]
```diff
@@ -84,7 +85,8 @@ SecretKeySpec getKey() {
     return outIv;
   }
 
-  private CryptoOutputStream createOutputStream(WritableByteChannel ch) throws IOException {
+  @VisibleForTesting
```
Comment (beliefer) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java]: You use a `VisibleForTesting` annotation to release modifier `private`, just for junit?
Maybe Encapsulation a top class used for junit is better?

File: common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java [hunk 2]
```diff
@@ -44,7 +44,8 @@
   @VisibleForTesting
   static final String ENCRYPTION_HANDLER_NAME = "TransportEncryption";
   private static final String DECRYPTION_HANDLER_NAME = "TransportDecryption";
-  private static final int STREAM_BUFFER_SIZE = 1024 * 32;
+  @VisibleForTesting
```
Comment (beliefer) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java]: I think `VisibleForTesting` annotation used for a member with modifier `final` is OK.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java (hunk 1 above)
Comment (zsxwing) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java]: This is package private. We usually use this to expose methods for junit. It's not worth to encapsulate a class just for unit tests.

**Comment by cloud-fan** [comment]
thanks, merging to 2.4/2.3!

File: common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java (hunk 1 above)
Comment (beliefer) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/TransportCipher.java]: OK, I get it.

**Pull request closed without merging**